File: leetcode/C++Solutions/merge-sorted-array.cpp
//class Solution {
//public:
    //void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        //int i = m - 1;
        //int j = n - 1;
        //int k = m + n - 1;
//
        //while(j >= 0){
            //if (i >= 0 && nums1[i] > nums2[j])
                //nums1[k--] = nums1[i--];
            //else
                //nums1[k--] = nums2[j--];
        //};
    //};
//};

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        
        int last = m + n - 1;

        while(m>0 && n>0){
            if(nums1[m-1] > nums2[n-1]){
                nums1[last] = nums1[m-1];
                m--;
            }
            else{
                nums1[last] = nums2[n-1];
                n--;
            }
            last--;
        }

        while(n>0){
            nums1[last] = nums2[n-1];
            n--;
            last--;
        }

    }
};

File: leetcode/pythonSolutions/merge-sorted-array.py
#class Solution:
    #def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        #"""
        #Do not return anything, modify nums1 in-place instead.
        #"""
        #nums1[m:m+n] = nums2
        #nums1.sort()
#

# The above solution is not an efficient solution, the below one is better comparitively

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """

        last = m + n - 1

        while n>0 and m>0:
            if nums1[m-1] < nums2[n-1]:
                nums1[last] = nums2[n-1]
                n -= 1
            else:
                nums1[last] = nums1[m-1]
                m -= 1
            last-=1

        while n>0:
            nums1[last] = nums2[n-1]
            n -= 1
            last -= 1



            
        
